sklearn/cluster/k_means_.py

```diff
@@ -920,6 +926,10 @@ def __init__(self, n_clusters=8, init='k-means++', n_init=10,
         self.copy_x = copy_x
         self.n_jobs = n_jobs
         self.algorithm = algorithm
+        if n_init == 'warn':
```

Comment:
The validation and raising should be done in `fit`.

You can check this PR:
https://github.com/scikit-learn/scikit-learn/pull/11469/files#diff-e6faf37b13574bc591afbf0536128735R858

which is not merged yet but shows how to do that in more details.